Issues/345 update structure

pboling opened the pull request:
# No functional runtime changes

Addresses #345 

- [x] Update to current RSpec syntax `RSpec.describe` at outset
- [x] Add `rubocop-rspec` for a suite of RSpec related rules
- [x] Readme updates
  - Update Copyright
  - Add development tips
  - Add contributing tips
  - Add Code of Conduct link
- [x] Update gemspec to define all development dependencies that are across all supported Rubies
- [x] Document the namespace of the library
- [x] Run specs in documentation format and auto-require `spec_helper` which now requires `helper`
  - remove `require 'helper'` lines, no longer needed.
- [x] Enhance `.gitignore` & document it
- [x] Add `CODE_OF_CONDUCT.md`
- [x] Add bin files: `bin/console` and `bin/setup`
- [x] Add new copyright line, current year, to `LICENSE.md`
- [x] Add more supported Rubies to build
  - `jruby-9.0`
  - `jruby-1.7`
  - `ruby-2.5`
- [x] Split out some Cops that we want to progressively adopt, but are currently disabled in `.rubocop_todo.yml`
- [x] Add `wwtd` gem to allow for local simulation of the travis build.

xtagon commented:
Alright, the changes look good to me. I especially like the Code of Conduct added. 👍

xtagon commented:
Any idea why the build has a rake error?

pboling commented:
@xtagon I need to do some additional gem version constraining by Ruby version.  I'll do that tonight.

josephpage commented on oauth2.gemspec:
```diff
@@ -10,17 +10,33 @@ Gem::Specification.new do |spec|
+  spec.bindir        = 'exe'
+  spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
+  spec.files         = `git ls-files -z`.split("\x0").reject do |f|
+    f.match(%r{^(test|spec|features)/})
```
The `bin` directory is probably for dev only and should be excluded too.

josephpage commented on .travis.yml:
```diff
@@ -11,21 +12,40 @@ env:
+#      gemfile: gemfiles/ruby_1.8.gemfile
+    - rvm: jruby-1.7.27
+      gemfile: gemfiles/jruby_1.7.gemfile
+#    - rvm: 1.9.3-p551
```
Do you think we should not try to support 1.9 & 2.0 ?

pboling replied:
No, they are only commented out for my local testing.  I will turn them back on once I get things stable.

pboling commented:
@xtagon This is done!  Mostly the same as when you reviewed.  Removed and ignored the gemfile.locks.  Added rubocop-rspec for more rules.  Turned back on the Rubies I had disabled.

xtagon commented:
Thanks @pboling
